docs: Handling High Traffic (#2199)
wmde/wbaas-deploy: doc/handling-high-traffic.md

# Handling High Traffic

## Introduction

> Starting in 2025 Wikibase cloud started occasionally receiving a large amount of traffic.

Reviewer: Is it also the type of traffic as well as the amount? I.e. querying pages that aren't cached and accept query parameters (e.g. recent changes).

This has caused a number of incidents:
- https://wmde.github.io/wikibase-cloud-status/issues/2025-02-10-nginx-oom/ / T385969

> - https://wmde.github.io/wikibase-cloud-status/issues/2025-07-15-scrapers-again/ / T399439 and T400046

Reviewer: Will GitHub automatically link the Phabricator tasks? If not, I think it would be nice to make them links.

Author: it does add nice links in commits but not in markdown

> It's wasted quite a lot of engineering time; if you come here after dealing with this in the future why not update below.

Reviewer: ```suggestion
It's wasted quite a lot of engineering time; if you come here after dealing with this in the future, please update below.
```

Wasted Engineering days: 26

> ## Solutions in place

Reviewer: We have also set the `$wgMaxExecutionTimeForExpensiveQueries` MediaWiki setting, which limits the amount of time MediaWiki can execute a database query for.

Author: good point I forgot to add this; done!

### robots.txt
We have a robots.txt in the [platform nginx](../k8s/helmfile/env/production/platform-nginx.nginx.conf). This helps prevent well behaved bot traffic from getting stuck in problematic parts of the site.

### rate limiting
These are set in the [ingress-nginx](../k8s/helmfile/env/production/ingress-nginx.values.yaml.gotmpl)

> #### rate limits per unique IP

Reviewer: ```suggestion

#### rate limits per unique IP
```

We rate limit per ip. That is for any single given IP we limit the rate of connections.

#### rate limits per site
We rate limit per Wikibase we host. That is there is a rate-limit for the number of connections to e.g. `coffeebase.wikibase.cloud` from any source.

#### rate limits from specific groups

> When we have seen bad traffic from a given IP range we may introduce shared limits for this IP range. Initially this was for a Data Center from a specific cloud provider that was generating wildly excess traffic.

Reviewer: Do you think it would be beneficial to include a link to the PR that made this change, in case people want to do this again in the future?

Author: Seems wise! Will do.
